proposals/0035-linalg-matrix.md

```diff
@@ -859,4 +923,216 @@ Stores a row-vector to a matrix. Out of bounds writes no-op.
+                     bool ColMajor, uint Align = sizeof(ElementType));
+  static Matrix Load(RWByteAddressBuffer Res, uint StartOffset, uint Stride,
+                     bool ColMajor, uint Align = sizeof(ElementType));
+
+  template <typename T>
+  static typename hlsl::enable_if<hlsl::is_arithmetic<T>::value, Matrix>::type
+  Load(/*groupshared*/ T Arr[], uint StartIdx, uint Stride, bool ColMajor);
+
+  void Store(RWByteAddressBuffer Res, uint StartOffset, uint Stride,
+             bool ColMajor, uint Align = sizeof(ElementType));
+
+  template <typename T>
+  typename hlsl::enable_if<hlsl::is_arithmetic<T>::value, void>::type
+  Store(/*groupshared*/ T Arr[], uint StartIdx, uint Stride, bool ColMajor);
+
+  // Row accesses
+  vector<ElementType, M> GetRow(uint Index);
+  void SetRow(vector<ElementType, M> V, uint Index);
+
+  // Element access
+  typename hlsl::enable_if<
+      __detail::ComponentTypeTraits<ComponentTy>::IsNativeScalar,
+      ElementType>::type
+  Get(uint2 Index);
+  typename hlsl::enable_if<
+      __detail::ComponentTypeTraits<ComponentTy>::IsNativeScalar, void>::type
+  Set(ElementType V, uint2 Index);
+
+  template <MatrixComponentType LHSTy, MatrixComponentType RHSTy, uint K>
+  typename hlsl::enable_if<
+      Use == MatrixUse::Accumulator && Scope == MatrixScope::Wave, void>::type
+  MultiplyAccumulate(const Matrix<LHSTy, M, K, MatrixUse::A, Scope>,
+                     const Matrix<RHSTy, K, N, MatrixUse::B, Scope>);
+
+  template <MatrixComponentType LHSTy, MatrixComponentType RHSTy, uint K>
+  typename hlsl::enable_if<
+      Use == MatrixUse::Accumulator && Scope == MatrixScope::Wave, void>::type
+  SumAccumulate(const Matrix<LHSTy, M, K, MatrixUse::A, Scope>,
+                const Matrix<RHSTy, K, N, MatrixUse::B, Scope>);
+
+  // Cooperative Vector outer product accumulate.
+  template <typename T>
+  typename hlsl::enable_if<Use == MatrixUse::Accumulator, void>::type
+  OuterProductAccumulate(const vector<T, M>, const vector<T, N>);
+};
+
+template <MatrixComponentType T, uint M, uint N, uint K>
+Matrix<T, M, N, MatrixUse::Accumulator, MatrixScope::Wave>
+Multiply(const Matrix<T, M, K, MatrixUse::A, MatrixScope::Wave>,
+         const Matrix<T, K, N, MatrixUse::B, MatrixScope::Wave>);
+
+// Cooperative Vector Replacement API
+// Cooperative Vector operates on per-thread vectors multiplying against B
+// matrices.
+
+template <typename OutputElTy, bool MatrixTranspose, typename InputElTy, uint M,
+          uint K, MatrixComponentType MatrixDT, MatrixScope Scope>
+vector<OutputElTy, K>
+Multiply(vector<InputElTy, M> InputVector,
+         Matrix<MatrixDT, M, K, MatrixUse::B, Scope> Matrix);
+
+template <typename OutputElTy, typename InputElTy, typename BiasElTy, uint M,
+          uint K, MatrixComponentType MatrixDT, MatrixScope Scope,
+          bool MatrixTranspose>
+vector<OutputElTy, K>
+MultiplyAdd(vector<InputElTy, M> InputVector,
+            Matrix<MatrixDT, M, K, MatrixUse::B, Scope> Matrix,

```

Comment:
nit: I'm assuming parameter names that are equal to types are alright according to the style rules?

Reply:
I don't think we have an explicit style rule, but I'll clean that up.